dvc/repo/diff.py

```diff
@@ -124,3 +129,33 @@ def _filter_missing(repo, paths):
             out = metadata.outs[0]
             if out.status().get(str(out)) == "not in cache":
                 yield path
+
+
+def _filter_diff_targets(repo, targets, old, new):
```

Comment:
@efiop do you think this filtering code belongs here or would it be better in `dvc/command/diff.py`?

Reply:
Being able to filter diffs via the internal API may be useful in the future, and is not really a UI specific issue, I think it makes sense to have the filtering code here in `repo/diff.py`.

Is there a particular reason you moved this code out into a separate function, rather than filtering by targets in `_paths_checksums()` like your original implementation? In DVC we may have repos containing millions of files that we need to iterate through in `diff()`. For performance reasons it's probably better for us to only do that iteration once, and filter out by target as we go in `_paths_checksums()`. But with this filtering code called after `_paths_checksums()`, we will end up iterating over the entire list of paths in the repo a second time.

Reply:
I moved it out of `_paths_checksums()` to report errors on any targets that do not exists (this matches `git diff` behaviour).

If this is not desirable let me know and I can move things back.

Reply:
So something like
```python
repo_tree = RepoTree(repo, stream=True)
missing_targets = {target for target in targets if not repo_tree.exists(target)}
```
could be used inside the `brancher` loop where we call `_paths_checksums(self)` to accomplish the same thing with much less performance overhead. At that point you only need to error out if there is any overlap between the missing target sets for both revisions after they've been retrieved.

Reply:
@pmrowla thanks for reviewing and the feedback!

I've incorporated your suggestions in https://github.com/iterative/dvc/pull/4938/commits/132636fc0198a5ef52b5c5e1922e81e9d7c951af.